File: crates/yokan/demo/links.py
# /// script
# requires-python = ">=3.14"
# ///
"""Models referencing models, gated: `Node | None` fields wire an
ownership chain in handlers, `Weak[Node]` is the not-owning back
pointer (it breaks the cycle, so dropping the root frees the chain
in both tiers at the same statement), and views read through
walrus-narrowed bindings.
"""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from yokan import button, column, model, row, run, State, store, text, Weak  # noqa: E402


@model
class Node:
    label: str = "n"
    kid: Node | None = None
    parent: Weak[Node] = None


@store
class Tree:
    root: Node | None = None
    keep: Node | None = None
    note: str = "-"

    def build(self) -> None:
        a = Node()
        a.label = "alpha"
        b = Node()
        b.label = "beta"
        a.kid = b
        b.parent = a
        self.root = a
        self.keep = b

    def drop_root(self) -> None:
        self.root = None

    def peek(self) -> None:
        if (r := Tree.root) is not None:
            if (k := r.kid) is not None:
                if (p := k.parent) is not None:
                    self.note = f"kid={k.label} parent={p.label}"
                else:
                    self.note = f"kid={k.label} parent=gone"
            else:
                self.note = "no kid"
        elif (k := Tree.keep) is not None:
            if (p := k.parent) is not None:
                self.note = f"kept {k.label}, parent={p.label}"
            else:
                self.note = f"kept {k.label}, parent=gone"
        else:
            self.note = "no root"


def view():
    with column(spacing=8, padding=12):
        text(f"note: {Tree.note}")
        if (r := Tree.root) is not None:
            text(f"root: {r.label}")
        else:
            text("root: (none)")
        with row(spacing=6):
            button("build", on_click=Tree.build)
            button("peek", on_click=Tree.peek)
            button("drop", on_click=Tree.drop_root)


if __name__ == "__main__":
    run(view, title="links")

File: rakugan/demo/links.pl
# Objects that point at one another. A perl object is a reference, and
# two names can hold the same one; the compiled run keeps that. The
# pointer back is weakened, as perl itself asks, so a parent and a
# child do not keep each other alive: cut the owning chain and the
# survivor's pointer back answers nothing, in both runs.
use Rakugan;

class Node {
    use Rakugan;
    field $label  :param :reader = "n";
    field $kid    :reader :writer = maybe(Node);
    field $parent :reader = maybe(Node);

    method hang_under :Sig(Node) ($p) {
        $parent = $p;
        weaken($parent);
    }
}

class Tree {
    use Rakugan;
    field $root = maybe(Node);
    field $keep = maybe(Node);
    field $note = "-";

    method build {
        my $a = Node->new(label => "alpha");
        my $b = Node->new(label => "beta");
        $a->set_kid($b);
        $b->hang_under($a);
        $root = $a;
        $keep = $b;
    }

    method peek {
        if (defined $root) {
            if (defined $root->kid) {
                if (defined $root->kid->parent) {
                    $note = "kid=@{[ $root->kid->label ]} parent=@{[ $root->kid->parent->label ]}";
                } else {
                    $note = "kid=@{[ $root->kid->label ]} parent=gone";
                }
            } else {
                $note = "no kid";
            }
        } elsif (defined $keep) {
            if (defined $keep->parent) {
                $note = "kept @{[ $keep->label ]}, parent=@{[ $keep->parent->label ]}";
            } else {
                $note = "kept @{[ $keep->label ]}, parent=gone";
            }
        } else {
            $note = "no root";
        }
    }

    method view {
        my @cells = (text("note: $note"));
        if (defined $root) {
            push @cells, text("root: @{[ $root->label ]}");
        } else {
            push @cells, text("root: (none)");
        }
        return column(
            @cells,
            row(
                button("build", on_click => sub { $self->build }),
                button("peek",  on_click => sub { $self->peek }),
                button("drop",  on_click => sub { $root = undef }),
                spacing => 6,
            ),
            spacing => 8,
            padding => 12,
        );
    }
}

run(Tree->new, title => "links");
